dll/ntdll/ldr/ldrinit.c

```diff
@@ -1538,7 +1538,36 @@ VOID
 NTAPI
 LdrpValidateImageForMp(IN PLDR_DATA_TABLE_ENTRY LdrDataTableEntry)
 {
-    UNIMPLEMENTED;
+    PIMAGE_NT_HEADERS NtHeaders;
+    NTSTATUS Status;
+
+    /* Get the NT Headers */
+    NtHeaders = RtlImageNtHeader(LdrDataTableEntry->DllBase);
+    if (NtHeaders == NULL)
+    {
+        RtlRaiseStatus(STATUS_INVALID_IMAGE_FORMAT);
+    }
+
+    /* Check if the image is UP only */
+    if (NtHeaders->FileHeader.Characteristics & IMAGE_FILE_UP_SYSTEM_ONLY)
+    {
+        /* Set process affinity to 1 */
+        DPRINT1("LDR: Image is UP only, setting process affinity to 1\n");
+        KAFFINITY AffinityMask = 1;
+        Status = NtSetInformationProcess(NtCurrentProcess(),
+                                         ProcessAffinityMask,
+                                         &AffinityMask,
+                                         sizeof(AffinityMask));
+        if (!NT_SUCCESS(Status))
+        {
+            RtlRaiseStatus(Status);
+        }
+
+        /* Save it in the PEB */
+        NtCurrentPeb()->NumberOfProcessors = 1;
+    }
+
+    // FIXME: Anything else to do?
```

Comment:
- Checking the image's `IMAGE_LOAD_CONFIG_DIRECTORY`, and perhaps do something with `ProcessAffinityMask` (if it's not already done by https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/mm/ARM3/procsup.c;hb=2da2fba042f013e5a21b9abe39d5709e95db81f6#l664 );
- Maybe also handle the [`LockPrefixTable`](https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-image_load_config_directory32) (as an optimization, on x86);
- Raise a `STATUS_IMAGE_MP_UP_MISMATCH` hard error if there's a problem?

Reply:
The referenced code does `Peb->ImageProcessAffinityMask = 0;` which is obviously a bug, but it turns out that the loader already sets the process affinity. I fixed `MmCreatePeb` instead and removed the obsolete code.